rfc/account-management.md

```diff
@@ -0,0 +1,135 @@
+# Spinnaker Account Management
+
+| | |
+|-|-|
+| **Status**    | _**Proposed**, Accepted, Implemented, Obsolete_       |
+| **RFC #**      | https://github.com/spinnaker/governance/pull/159 |
+| **Author(s)**  | Nicolas Cohen (@ncknt), Nima Kaviani (@nimakaviani) |
+| **SIG / WG**   | Platform |
+
+## Overview
+
+Spinnaker reads account definitions (cloud providers, CI, metric stores, etc.) from Spring properties. This works well in a world with a handful of accounts that rarely change but causes operational pain when provisioning accounts dynamically or with account information stored externally.
+
+This document proposes to add a thin account management layer in each Spinnaker service to support dynamic account management from non Spring data sources.
+
+
+### Goals and Non-Goals
+
+Goals:
+- Provide an interface that each Spinnaker service can optionally implement
+- Support Spring properties (as today)
+- Adding account dynamically without having to restart the service
+- Add stable interfaces that can pull accounts from a custom source
+- Prepare Clouddriver for multiple credential repositories
+
+Non Goals:
+- Storing account information in Spinnaker
+- Migrate all services and all providers
+
+
+## Motivation and Rationale
+
+Spinnaker reads account definitions (cloud providers, CI, metric stores, etc.) from Spring properties. This forces Spinnaker operators to restart services when an account is added, modified, or removed. The community is currently using [Spring Cloud Config](https://spinnaker.io/setup/configuration/) but it is complicated to set up and has significant performance issues as the number of accounts grow.
+
+We're proposing to define interfaces to let each Spinnaker service support a variety of sources of account definitions as well as dynamically reload them.  We'll work in each service to add support - one provider at a time.
+
+With this proposal implemented, operators of Spinnaker could easily:
+- reload accounts dynamically across all types of accounts
+- add a new account source from a specific provider
+- extend the behavior of the `CredentialsRepository` across all types of accounts.
+- add a new account source and get the account information from its source of truth.
+
+We have evidence from our customers that some form of dynamic account management is needed.
+
+## Timeline
+
+Spinnaker 1.23 (1.22?):
+- Add interfaces to kork 
+- Start splitting AccountCredentialsRepository by provider (we'd now be able to use multiple accounts with the same name in different providers)
```

Comment:
Is that preferable? We've got a weird mix of APIs that sometimes include account + provider and sometimes just account.

Historically I've come down on the side of appreciating uniqueness in account names (but I could _possibly_ be swayed here).

Reply:
I couldn't spot where the account was ambiguous if you know the provider (or the implied provider based on where you are in Deck or in the code). It clearly doesn't mean there aren't any!

I think it is important to allow different providers to use different sources. For instance, some users store their Kubernetes account information in a separate system but have a handful of manually configured AWS accounts that remain in Spring props.

The non uniqueness is a by-product of having separate sources of accounts. Currently, I am not even sure what happens when the same name is used in the different providers. The last one wins? And it probably changes if accounts are dynamic. Regardless, it's probably not what the user expects.

IMO we should make sure that the provider is always displayed next to the account (in Deck) or present in the URL in APIs.

Reply:
I didn't think `clouddriver` would even allow you to have duplicate account names across providers (we have two cloud providers that ultimately land resources in the same aws account but we give them distinct names.

Reply:
I believe you can configure it that way (in the yaml files) and I don't believe you'd get an error, just unexpected behavior.